app/helpers/protocols_io_helper.rb

```diff
@@ -145,4 +145,283 @@ def not_null(attribute)
+        ProtocolsIoHelper::PIO_ELEMENT_RESERVED_LENGTH_SMALL
+      )
+    when 'medium'
+      pio_eval_len(
+        sanitize_input(string_html_table_remove(not_null(attribute_text))),
+        ProtocolsIoHelper::PIO_ELEMENT_RESERVED_LENGTH_MEDIUM
+      )
+    when 'big'
+      pio_eval_len(
+        sanitize_input(string_html_table_remove(not_null(attribute_text))),
+        ProtocolsIoHelper::PIO_ELEMENT_RESERVED_LENGTH_BIG
+      )
+    when 'small-no-table'
+      pio_eval_len(
+        sanitize_input(not_null(attribute_text)),
+        ProtocolsIoHelper::PIO_ELEMENT_RESERVED_LENGTH_SMALL
+      )
+    when 'medium-no-table'
+      pio_eval_len(
+        sanitize_input(not_null(attribute_text)),
+        ProtocolsIoHelper::PIO_ELEMENT_RESERVED_LENGTH_MEDIUM
+      )
+    when 'big-no-table'
+      pio_eval_len(
+        sanitize_input(not_null(attribute_text)),
+        ProtocolsIoHelper::PIO_ELEMENT_RESERVED_LENGTH_BIG
+      )
+    end
+  end
+
+  def fill_attributes(attribute_name, attribute_text, step_component_key)
+    output_string = ''
+    trans_string = ''
+    case step_component_key
+    when '8'
+      trans_string = 'protocols.protocols_io_import.comp_append.soft_packg.'
+    when '9'
+      trans_string = 'protocols.protocols_io_import.comp_append.dataset.'
+    when '15'
+      trans_string = 'protocols.protocols_io_import.comp_append.command.'
+    when '18'
+      trans_string = 'protocols.protocols_io_import.comp_append.sub_protocol.'
+    when '19'
+      trans_string = 'protocols.protocols_io_import.comp_append.safety_infor.'
+    end
+    trans_string +=
+      if attribute_name != 'os_name' && attribute_name != 'os_version'
+        attribute_name
+      else
+        'os'
+      end
+    output_string +=
+      if attribute_name != 'os_version'
+        t(trans_string)
+      else
+        ' , '
+      end
+    output_string += prepare_for_view(attribute_text, 'small-no-table')
+    output_string
+  end
+
+  # pio_stp_x means protocols io step (id of component) parser
+  def pio_stp_1(iterating_key) # protocols io description parser
+    br = '<br>'
+    append =
+      if iterating_key.present?
+        br +
+        prepare_for_view(iterating_key, 'small-no-table') +
+        br
+      else
+        t('protocols.protocols_io_import.comp_append.missing_desc')
+      end
+    append
+  end
+
+  def pio_stp_6(iterating_key) # protocols io section(title) parser
+    return pio_eval_title_len(sanitize_input(iterating_key)) if iterating_key.present?
+    t('protocols.protocols_io_import.comp_append.missing_step')
+  end
+
+  def pio_stp_17(iterating_key) # protocols io expected result parser
+    if iterating_key.present?
+      append =
+        t('protocols.protocols_io_import.comp_append.expected_result') +
+        prepare_for_view(iterating_key, 'small-no-table') +
+        '<br>'
+      return append
+    end
+    ''
+  end
+
+  def pio_stp_8(iterating_key) # protocols io software package parser
```

Comment:
You could move all the if checks inside the `parse_elements_array` loop like this:
```
parse_elements_array.each do |element|
  return ' ' unless iterating_key[element]
  append += fill_attributes(element, iterating_key[element], '8')
end
```
You can do these changes also in other similar methods:` pio_stp_9(iterating_key)`, `pio_stp_15(iterating_key)`, itd.